numpy/core/src/multiarray/convert_datatype.c

```diff
@@ -1927,17 +1927,26 @@ PyArray_Zero(PyArrayObject *arr)
         return NULL;
     }
 
-    obj=PyInt_FromLong((long) 0);
+    if (zero_obj == NULL) {
+        zero_obj = PyInt_FromLong((long) 0);
+        if (zero_obj == NULL) {
+            return NULL;
+        }
+    }
     if (PyArray_ISOBJECT(arr)) {
```

Comment:
My analysis was wrong, yours was more correct. But in the end, we cannot use ``setitem``, when I tried I got a segfault in ``test_simple_object`` from ``test_multiarray``.

Theoretically ``PyArray_DESCR(arr)->f->setitem(one_obj, oneval, arr)``, which ends up on ``OBJECT_setitem``, should work, although it would call ``Py_INCREF`` on the ``PyObject*`` so we would need a matching ``Py_DECREF`` after the call, who would do it and when?

Unfortunately the current implementation crashes since ``OBJECT_setitem`` calls ``Py_XDECREF`` on ``oneval`` (or ``zeroval``) and since the memory was allocated with ``malloc`` via ``PyDataMem_New`` it is not zeroed out, so the ``Py_XDECREF`` crashes the interpreter. Perhaps this is why the original implementer did not call ``setitem``, in addition to the concern about the extra ``Py_INCREF`` needing a matching call somewhere.